docs/design/2018-07-22-enhance-propagations.md

```diff
@@ -0,0 +1,189 @@
+| │   └─TableScan_8       | 2.00  | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo             |
+| └─TableReader_13        | 0.00  | root | data:Selection_12                                                       |
+|   └─Selection_12        | 0.00  | cop  | in(test.t2.a, 12, 13)                                                   |
+|     └─TableScan_11      | 2.00  | cop  | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo             |
++-------------------------+-------+------+-------------------------------------------------------------------------+
+7 rows in set (0.00 sec)
+```
+
+## Rationale
+
+Constraint propagation is commonly used as logcial plan optimization in traditional databases, for example, [this doc](https://dev.mysql.com/doc/internals/en/optimizer-constant-propagation.html) explained some details of constant propagtions in MySQL. It is is also widely adopted in distributed analytical engines, like Apache Hive, Apache SparkSQL, Apache Impala, et al, those engines usually query on huge amount of data.
+
+### Advantages:
+
+ Constraint propagation brings more detailed, explicit constraints to each data source involved in a query. With those constraints we can filter data as early as possible, and thus reduce disc/network IO and computational overheads during the execution of a query. In TiDB, most propagated filters can be pushed down to the storage level(TiKV) as a coprocessor task, and lead to the following benefits:
+
+ * Apply the filters at each TiKV instance, which make the calculation distributed.
+
+ * When loading data, skip some table partitions if its data range doesn't pass the filter
+
+ * For the columnar storage format to be supported in the future, we may apply some filters directly when accessing the raw storage.
+
+ * Reduce the data transfered from TiKV to TiDB
+
+### Disadvantages:
+
+ Constraint propagation may brings unnecessary filters and leads to unnecessary overheads during a query, this is mostly due to the fact that logical optimization doesn't take data statistics into account, for example:
+
+ For a query `select * from t0, t1 on t0.a = t1.a where t1.a < 5`, we get a propagation `t0.a < 5`, but if all `t0.a` is greater than 5, applying the filter brings unnecessary overheads.
+
+Considering the trade-off, most of the time we gain benefits from constraint propagation and still treat it useful.
+
+## Compatibility
+
+All rules mentioned in this proposal are logical plan optimization, they should not change the semantic of a query, and thus dont't lead to any compatibility issue
+
+## Implementation
+
+Here are rough ideas about possible implementations:
+
+ * For proposal #1, we can extend current `propagateConstantSolver` to support wider types of operators from column equiality.
+
+ * For proposal #2, `propagateConstantSolver` is also a applicable way to add `NotNULL` filter(`not(isnull())`), but should examine if the column doesn't have `NotNULL` constraint.
```

Comment:
"For proposal #2, `propagateConstantSolver` is also **an** applicable way to add `NotNULL` filter(`not(isnull())`), but should examine **whether** the column has **the** `NotNULL` constraint"?
or
"For proposal #2, `propagateConstantSolver` is also **an** applicable way to add `NotNULL` filter(`not(isnull())`), but should **be examined** if the column doesn't have **the** `NotNULL` constraint"?